File: binary-tree-in-python/perfect-binary-tree.py
# create Node
class Node:
    def __init__(self, data):
        self.data = data
        self.left = None
        self.right = None

# search depth
def depth(root):
    temp = root
    i = 0
    while temp is not None:
        temp = temp.left
        i += 1
    return i

# check full binary tree
def isPbt(root, depth, level):
    if root is None:
        return True
    if (root.left is None) and (root.right is None):
        return depth == level+1
    if root.left and root.right:
        return isPbt(root.left, depth, level+1) and isPbt(root.right, depth, level+1)
    return False


if __name__ == '__main__':
    root = Node(1)
    root.left = Node(2)
    root.right = Node(3)
    root.left.left = Node(4)
    root.left.right = Node(5)
    root.right.left = Node(6)
    root.right.right = Node(7)
    if isPbt(root, depth(root), 0):
        print("This is perfect binary tree")
    else:
        print("This is not perfect binary tree")

File: binary-tree-in-C++/perfect-binary-tree.cpp
#include <iostream>
using namespace std;

// create Node
class Node{
public:
    int data;
    Node *left, *right;

    Node(){
        this->left = this->right = NULL;
    }
};
// search depth
int depth(Node* root){
    Node* temp = root;
    int i=0;
    while(temp != NULL){
        temp = temp->left;
        i++;
    }
    return i;
}

//test perfect-binary-tree
bool isPbt(Node* root, int depth, int level){
    if(root == NULL){
        return true;
    }
    if(root->left == NULL && root->right == NULL){
        return depth == level+1;
    }
    if(root->left && root->right){
        return isPbt(root->left, depth, 1+level) && isPbt(root->right, depth, 1+level);
    }
    return false;
}

int main(){
    Node *root = new Node();
    root->data = 1;
    root->left = new Node();
    root->left->data = 2;
    root->right = new Node();
    root->right->data = 3;
    root->left->left = new Node();
    root->left->left->data = 4;
    root->left->right = new Node();
    root->left->right->data = 5;
    root->right->left = new Node();
    root->right->left->data = 6;
    root->right->right = new Node();
    root->right->right->data = 7;

    bool is_perfect_binary_tree = isPbt(root, depth(root), 0);
    if(is_perfect_binary_tree){
        cout << "this is perfect binary tree" << endl;
    }else{
        cout << "this is not perfect binary tree" << endl;
    }
    return 1;
}